File: Kiro/go/exceptions.go
package gridneighborhoods

import "fmt"

// InvalidGridDimensionsError represents an error when grid dimensions are invalid
type InvalidGridDimensionsError struct {
	Height int
	Width  int
}

func (e *InvalidGridDimensionsError) Error() string {
	return fmt.Sprintf("invalid grid dimensions: height=%d, width=%d (both must be > 0)", e.Height, e.Width)
}

// PositionOutOfBoundsError represents an error when a position is outside grid boundaries
type PositionOutOfBoundsError struct {
	Position Position
	Height   int
	Width    int
}

func (e *PositionOutOfBoundsError) Error() string {
	return fmt.Sprintf("position (%d,%d) is out of bounds for grid %dx%d", e.Position.Row, e.Position.Column, e.Height, e.Width)
}

// InvalidDistanceThresholdError represents an error when distance threshold is negative
type InvalidDistanceThresholdError struct {
	Threshold int
}

func (e *InvalidDistanceThresholdError) Error() string {
	return fmt.Sprintf("invalid distance threshold: %d (must be >= 0)", e.Threshold)
}

File: Kiro/csharp/src/Exceptions.cs
namespace GridNeighborhoods;

/// <summary>Base type for grid neighborhood errors.</summary>
public abstract class GridNeighborhoodsException : Exception
{
    protected GridNeighborhoodsException(string message) : base(message) { }
}

/// <summary>
/// Thrown when a grid is created with height &lt;= 0 or width &lt;= 0.
/// Requirement 8.1.
/// </summary>
public sealed class InvalidGridDimensionsException : GridNeighborhoodsException
{
    public InvalidGridDimensionsException(long height, long width)
        : base($"invalid grid dimensions: height={height} width={width} (both must be > 0)") { }
}

/// <summary>
/// Thrown when a positive cell position falls outside the grid boundaries.
/// Requirement 8.2.
/// </summary>
public sealed class PositionOutOfBoundsException : GridNeighborhoodsException
{
    public PositionOutOfBoundsException(Position position, long height, long width)
        : base($"position ({position.Row}, {position.Column}) is out of bounds for a {height}x{width} grid") { }
}

/// <summary>
/// Thrown when the distance threshold N is negative.
/// Requirement 8.3.
/// </summary>
public sealed class InvalidDistanceThresholdException : GridNeighborhoodsException
{
    public InvalidDistanceThresholdException(long threshold)
        : base($"invalid distance threshold: {threshold} (must be >= 0)") { }
}
